File: src/sorting/treeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "./_common_/common.h"

// Make a simplified binary search tree

typedef struct _node
{
    int data;
    struct _node* left;
    struct _node* right;
    int count;
} Node;

Node* get_node(const int d)
{
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = d;
    newNode->left = newNode->right = NULL;
    newNode->count = 1;
    return newNode;
}

Node* insert(Node* root, const int d) // Put data into tree.
{
    if (!root)
    {
        root = get_node(d);
        return root;
    }

    if (root->data > d) root->left = insert(root->left, d);
    else if (root->data < d) root->right = insert(root->right, d);
    else root->count++;
    return root;
}

void print_tree(Node* root) // Print all elements in the binary search tree.
{
    if (!root) return;

    print_tree(root->left);
    for (int i = 0; i < root->count; i++)
        printf("%d ", root->data);
    print_tree(root->right);
    return;
}

void free_all(Node* root) // Deallocate memory before the end of program.
{
    if (!root) return;

    if (root->left) free_all(root->left);
    if (root->right) free_all(root->right);
    free(root);
    return;
}

int main(void)
{
    int n;
    Node* t = NULL;

    scanf("%d", &n);
    int* arr = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &arr[i]);
        t = insert(t, arr[i]);
    }
    print_array(arr, n);
    print_tree(t);

    free_all(t);
    free(arr);
    return 0;
}

/*
Input:
5
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
*/

File: src/sorting/treeSort.cs
using System;
using Sorting;

namespace TreeSort
{
    public class TreeSort<T> : Sorting<T> where T : IComparable<T>
    {
        public TreeSort()
        {
            _arr = Array.ConvertAll<string, T>(Console.ReadLine().Split(' ',
            StringSplitOptions.RemoveEmptyEntries), ParseType);
            _count = _arr.Length;
        }

        public TreeSort(T[] input)
        {
            _arr = input;
            _count = _arr.Length;
        }

        internal class Node
        {
            private T _data;

            public T Data { get { return _data; } }
            public Node? Left { get; set; }
            public Node? Right { get; set; }
            public int Count { get; set; }

            public Node(T val)
            {
                _data = val;
                Left = Right = null;
                Count = 0;
            }
        }

        private Node? InsertNode(Node? root, T val)
        {
            if (root == null)
            {
                root = new Node(val);
                return root;
            }

            if (val.CompareTo(root.Data) < 0)
                root.Left = InsertNode(root.Left, val);
            else if (val.CompareTo(root.Data) > 0)
                root.Right = InsertNode(root.Right, val);
            else root.Count++;

            return root;
        }

        private int Inorder(Node? root, int idx)
        {
            if (root == null) return idx;

            if (root.Left != null) idx = Inorder(root.Left, idx);
            _arr[idx++] = root.Data;
            if (root.Right != null) idx = Inorder(root.Right, idx);

            return idx;
        }

        public override void Sort()
        {
            Node? rt = null;
            foreach (var item in _arr)
                rt = InsertNode(rt, item);
            Inorder(rt, 0);
        }
    }

    public static class Program
    {
        public static void Main()
        {
            var a = new TreeSort<int>();
            a.PrintArray();
            a.Sort();
            a.PrintArray();
        }
    }
}